### DISCUSSION TITLE
Something I do not understand about Grid Layout Auto

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#16918] [Q&A] [answered]
Hi, 
I am using Grids for 15 years with no trouble but today I am faced to a behavior that I don't understand... 
Initially I believed it was a bug, but I tested in WPF and the layout is exactly the same!

Here is the code:
```
<Grid
    x:Name="myGrid"
    VerticalAlignment="Top"
    Background="Red"
    ColumnDefinitions="*,Auto"
    ShowGridLines="True">
    <Grid.RowDefinitions>
        <RowDefinition Height="6" />
        <RowDefinition Height="Auto" />
        <RowDefinition Height="6" />
    </Grid.RowDefinitions>

    <Rectangle
        x:Name="SliderTrack"
        Grid.Row="1"
        Height="20"
        Fill="Pink" />

    <local:Button
        Grid.Row="0"
        Grid.RowSpan="3"
        Grid.Column="1"
        Height="32"
        Margin="8,0,0,0"
        Background="Yellow"
        Content="Button" />
</Grid>
<TextBlock Text="{Binding Bounds.Height, ElementName=myGrid}" />
``` 
And here is the rendering:
![image](https://github.com/user-attachments/assets/5ca8139a-2d18-4929-b400-3eb91261ae1f)

Basically I have a Grid with 3 rows of sizes 6, Auto, 6.
The middle row contains a control if height = 20, so the total height is 32.
Now my problem, I add another Control of height = 32 in the last column with RowSpan=3... 

To me, the total height of the Grid should be 32... as the button is span over 3 rows... but it is 52 !

If I force the height of the middle row to 20 or if I use "*", I got the result I am expecting... But to me, Auto should work too, and take the size of the Rectangle... 

What I am missing?!

CORRECTION:
> If I force the height of the middle row to 20 or if I use "*", I got the result I am expecting... But to me, Auto should work too, and take the size of the Rectangle...

If I set the middle row to "*", it work when the control is in a grid, but not when the control is in a StackPanel?!?! (same in WPF)

Changing the middle row to "*":
![image](https://github.com/user-attachments/assets/3cf481e9-729f-4226-8c5c-7ca81d0c9321)

Enclosing the control in a StackPanel result in :
![image](https://github.com/user-attachments/assets/dc46fd36-14d7-4fb2-aa05-e620af676c32)

???

**Comment by dbriard** [comment]
Hum, if I use two grids, one for the columns, and one for the rows, everything work as I expect... but I do not understand the difference with a single grid with rows and columns... if someone could explain?

![image](https://github.com/user-attachments/assets/805b5b88-2357-4d4d-8a7f-6f4b3a93d75b)
![image](https://github.com/user-attachments/assets/45b3f55d-c09e-430d-96e7-f213a4a11fec)

**Comment by IanRawley** [comment, marked as answer]
The algorithm for calculating Grid Cell sizes is pretty complicated, as you can see [here:](https://github.com/AvaloniaUI/Avalonia/blob/77fe0400fc57139c4a53fc0c7d148fca36c80cdd/src/Avalonia.Controls/Grid.cs#L277-L425) 

From what I can figure out, the Grid arranges its children into Cells.  In this case there are two children, the Rectangle and the Button.  
The cells are then arranged into groups based on their Row/Column definitions.  The Button's span falls into group 1, so is calculated first.  It requests a size of 32 for the span (the height of the Button), but the span has a preferred size of 12 based on the RowDefinitions (6 + 0(Auto) + 6).  It needs 20 more pixels, so it divides those between the rows that aren't Auto, Rows 0 and 2, updating their minimum sizes.  So Row 0 and Row 2 (which want a size of 6) are assigned a minimum size of 16, while the auto row has a minimum and preferred of 0 still.

It then calculates the size of the cell for the rectangle.  The height is set to 20, so that's easy, making the preferred row height 20 for that row.  All the cells have now had their height calculated and RowDefinition Minimum Height and Preferred height properties adjusted, so the final heights are 16 (row 0 minimum), 20(row 1 preferred), 16(row 2 minimum) for a total of 52 pixels of height.

**Reply by IanRawley** [reply]
As a result, the layout in the example actually boils down to a single row 2 column grid, and if done as such with the fixed heights (and rectangle centred vertically) would probably function as expected.  The deep-dive has been eye opening to me, and if I have a precise image of the final layout in mind I'll be avoiding Row, Column positions without controls assigned where ever possible going forward as they can make the final layout unpredictable.

**Reply by MarkusKgit** [reply]
That was very insightful 👍
Btw it works like intended if you also also set the `MaxHeight` to 6 on the top and bottom `RowDefinition`. I guess the automatic layout always respects that whereas the Height is more of a request.

**Reply by dbriard** [reply]
Thanks for the detailed explanation @IanRawley , and for the tips @MarkusKgit 
I didn't expected that the height of the first and last rows could increase as I set Height="6" (fixed height to me).

I have the feeling I used that kind of layout without with no trouble for years, but probably with "*" instead of "Auto", subtle difference.

**Reply by IanRawley** [reply]
The biggest part for me in digging into this was finding that references to a "Cell" in the code weren't a cell in the same way you'd see them in a spreadsheet, they're effectively a wrapper around a control.  That combined with Height / Width sizes being more polite requests than concrete sizes and humans being very good at spotting patterns makes the end result wildly different than expected.

For * vs. Auto on the middle row in this case, the big difference is that when assigning space for the span the middle row will grow as well.  The algorithm wants Auto sizes to be as tight as possible, presumably because if there is nothing actually in the row the height should be zero, and that's why the span assigns extra space to other rows.  If you make that * instead, that constraint disappears and the 20 extra pixels of space will be distributed over all 3 rows equally.  The result there would then be something similar to 12.66, 6.66, 12.66 minimum sizes, which gets turned into 12.66, 20, 12.66 for a total of about 45.3 as seen in the stack panel picture with some slight rounding differences.  When placed in a Grid, the external grid applies constraints to its children, so if the external grid is set to a specific height then the children's max height is capped based on the external Grid's settings.

**Reply by dbriard** [reply]
@IanRawley Congrats on the research and thank you so much, I learned a lot :) I cannot reproduce the different size with the StackPanel today (I probably changed something) but your explanation make sense.
